Add initial binary package for a Sway session

kgilmer opened the pull request:
Here is a sub for a new Sway session.  This represents the top level depencency for the session.  As you may notice when comparing to the flashback session, at this point in the package graph there is very little regolith packages referenced.  So there is a lot of flexibility.

kgilmer committed:
Add initial binary package for a Sway session

SoumyaRanjanPatnaik committed:
Initial wayland setup

SoumyaRanjanPatnaik committed:
Updated control file, minor changes to script

SoumyaRanjanPatnaik committed:
Updated wayland session file

SoumyaRanjanPatnaik committed:
Update session name and config location

SoumyaRanjanPatnaik committed:
Updated regolith-look for wayland support

SoumyaRanjanPatnaik committed:
Added install file for wayland binary package

SoumyaRanjanPatnaik commented:
Ready to merge

kgilmer commented on debian/control:
```diff
@@ -30,3 +30,22 @@ Description: Regolith customized gnome-flashback session
+Package: regolith-session-sway
+Architecture: any
+Depends: ${misc:Depends},
+    dbus # add more dependencies here,
```
correctness: This line will fail to parse, can you remove the comment?

kgilmer commented on debian/control:
```diff
@@ -30,3 +30,22 @@ Description: Regolith customized gnome-flashback session
+    xwayland,
+    clipman,
+    wl-clipboard,
+    #trawld,
```
correctness: I believe this will fail to parse because due to the comment, there is nothing after `wl-clipboard` to read.  Either remove the comma on line 42 or enable the commented out lines.

SoumyaRanjanPatnaik replied:
Actually I intend to have those lines enabled. But for now these packages don't really exist😅. I'll probably remove these for now and add them back once everything is packaged

kgilmer commented on usr/bin/regolith-diagnostic:
```diff
@@ -47,3 +47,6 @@ gsettings list-recursively org.regolith-linux.ilia
 gsettings list-recursively org.gnome.desktop.wm.preferences
+
+echo "--- trawl"
+trawl -q
```
nice!

kgilmer commented on usr/bin/regolith-session-init:
```diff
@@ -5,28 +5,47 @@ set -Eeu -o pipefail
-if [ -n "$DESKTOP_AUTOSTART_ID" ]; then
-    dbus-send --print-reply --session --dest=org.gnome.SessionManager "/org/gnome/SessionManager" org.gnome.SessionManager.RegisterClient "string:Regolith" "string:$DESKTOP_AUTOSTART_ID"
-fi
+start_i3 () {
```
I'm fine to merge this as is but wonder if you considered breaking out the wayland specific login into a separate file.  IMO this would be a better approach as it scales better.  EG if I make a change to this file as-is I may have to test both x11 and sway, whereas if they were discrete files I wouldn't.  Is there a reason why breaking the logic out is problematic?

SoumyaRanjanPatnaik replied:
I thought it would be convenient to have the same starter script for both sway and i3. But on second thought, what you're saying does make a lot of sense.

kgilmer commented on usr/bin/regolith-session-init:
```diff
@@ -5,28 +5,47 @@ set -Eeu -o pipefail
-if [ -n "$DESKTOP_AUTOSTART_ID" ]; then
-    dbus-send --print-reply --session --dest=org.gnome.SessionManager "/org/gnome/SessionManager" org.gnome.SessionManager.Logout "uint32:1"
+export MOZ_ENABLE_WAYLAND=1
+export STUDIO_JDK=/usr/lib/jvm/java-11-openjdk/
```
Hmm..what is this?

SoumyaRanjanPatnaik replied:
Some applications require additional environment variables to work well with Wayland. It's quite common to have such variable definitions included in starter scripts. And these variables need to be declared before sway starts.

SoumyaRanjanPatnaik replied:
The second line is necessary for a lot of Java based applications (especially for Intellij based IDEs) to work properly.

SoumyaRanjanPatnaik replied:
Although, I think it might be a good idea to let users define their pre-startup environment variables at one place. What are your thoughts on this?

kgilmer replied:
My opinion is app specific logic should not live in general-purpose scripts.  For a user that does not use Firefox (assuming that from `Moz`) or care about java programs would classify this as bloat IMO.  Also, what's to say that the JVM path is the one a user really wants Inteliij to point to, or is even valid?

The simplest thing we can do is provide guidance/docs for poorly behaving apps that users encounter.  More work would be to create packages that install support for some applications like these.  I'm guessing there is something in Wayland like `/etc/X11/Xsession` that allows the user environment to be extended by a package.

kgilmer commented on usr/bin/regolith-session-init:
```diff
@@ -5,28 +5,47 @@ set -Eeu -o pipefail
-# Close session when i3 exits.
-if [ -n "$DESKTOP_AUTOSTART_ID" ]; then
-    dbus-send --print-reply --session --dest=org.gnome.SessionManager "/org/gnome/SessionManager" org.gnome.SessionManager.Logout "uint32:1"
+export MOZ_ENABLE_WAYLAND=1
```
Why is this needed?

SoumyaRanjanPatnaik replied:
In my experience, without this Firefox will use legacy playback methods (as in hardware acceleration won't work). This leads to unacceptable CPU usage and occasional jittering when playing in 4k.

SoumyaRanjanPatnaik replied:
There's also the added benefit of support for touchpad gestures (like pinch to zoom) and maybe a few other features that can only work in Wayland.

kgilmer replied:
As mentioned, I think this belongs elsewhere or possibly just in documentation.

kgilmer commented on usr/share/applications/regolith.desktop:
```diff
@@ -2,7 +2,7 @@
 Name=regolith
 Comment=Productivity-focused desktop environment, Regolith
-Exec=regolith-session-init
+Exec=regolith-session-init i3
```
interesting strategy.  I would there be a problem if we simply execute separate init scrips for each backend rather than having the script handle both?

SoumyaRanjanPatnaik replied:
Actually no. Having separate scripts would work just the same.

kgilmer commented on usr/lib/regolith/regolith-session-common.sh:
```diff
@@ -65,3 +84,22 @@ load_regolith_xres() {
+        trawldb --merge "$USER_XRESOURCE_OVERRIDE_FILE" 
+    fi
+
+    LOOK_STYLE_ROOT_PATH=$(trawlcat regolith.look.path "$DEFAULT_XRESOURCE_LOOK_PATH")
```
nice!

kgilmer reviewed (commented):
A few questions and fixes needed to debian packaging data.  Great work!

SoumyaRanjanPatnaik committed:
Streamlined regolith-look experience

SoumyaRanjanPatnaik committed:
Control file fixes

SoumyaRanjanPatnaik committed:
Restructured regolith-session into separate file for i3 and sway

CHANGELOG:
1. Moved logic for initializing sway session into 'regolith-session-sway'
2. Added option to export environment variables set by user
3. Added example user environment file (/etc/regolith/sway/environment)
4. Updated session files (/usr/share/application/regolith.desktop) and (/usr/share/wayland-sessions/regolith-sway.desktop)
5. Updated regolith-session-sway.install file

SoumyaRanjanPatnaik commented:
I've made some changes based on your comment. Can you review them?

SoumyaRanjanPatnaik committed:
Overwrite cleaup file instead of appending to it

kgilmer commented on etc/regolith/sway/environment:
```diff
@@ -0,0 +1,5 @@
+# For using firefox with wayland
```
Kindly remove these for now

kgilmer replied:
I'd like to begin testing this session from a built package so we'll address these app specific configs later.

kgilmer commented on usr/bin/regolith-session-sway:
```diff
@@ -0,0 +1,29 @@
+resolve_default_config_file
+
+# Preloads user environment variables
+USER_ENVIRONMENT=${trawlcat sway.preload_env $HOME/.config/regolith2/sway/environment}
```
I would prefer if this could be expressed in a standard way if possible.

https://wiki.gnome.org/Initiatives/Wayland/SessionStart

https://github.com/sddm/sddm/issues/916

https://superuser.com/questions/992210/how-to-set-environment-variables-for-a-gnome-wayland-session

Hmm, seems like maybe Wayland doesn't provide anything (yet)..?  Are you aware of anything that my 5 mins of google searching didn't uncover?  If not let's keep this logic but remove the default that has the moz and jdk stuff.

kgilmer commented on usr/bin/regolith-session-sway:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+# This script initializes Regolith Xresources and launches i3.
```
copypasta

SoumyaRanjanPatnaik committed:
Register sway with gnome-session
